### DISCUSSION TITLE
Memray not running on distroless

### DISCUSSION THREAD

**Discussion by Partho** [bloomberg/memray#664] [Q&A] [answered]
Hi team,
     I am copying python libraries from base image to distroless, however the distroless is giving this error while I am trying to run it:
```shell
 File "/home/dsci/classifier-api/app/core/middleware.py", line 13, in <module>
    import memray
  File "/venv/lib/python3.9/site-packages/memray/__init__.py", line 2, in <module>
    from ._memray import AllocationRecord
ImportError: cannot import name 'AllocationRecord' from 'memray._memray' (unknown location)
``` 

My base image(referred as `build-parent` below) already contains memray in the location `/venv/lib/python3.9/site-packages` . Also, my distroless settings look like this:
```Dockerfile
FROM google-gcr.artifactory.cicd.dc/distroless/python3-debian11 AS runtime

COPY --from=build-parent /venv /venv
COPY --from=build-parent /usr/lib/x86_64-linux-gnu /usr/lib/x86_64-linux-gnu
COPY --from=build-parent /home/dsci /home/dsci
COPY --from=build-parent /usr/local/lib/python3.9/dist-packages /venv/lib/python3.9/site-packages

WORKDIR /home/dsci/classifier-api/app
ENTRYPOINT ["/venv/bin/python", "entrypoint.py"]
``` 

Both base image and distroless image are based on debian11 and uses python3.9. Any idea what might be the issue and how to resolve it (library imports, path variables etc.)?

P.S. I see that AllocationRecord in present as a struct in `memray/_memray/records.h` (in distroless):
```cpp
struct AllocationRecord
{
    uintptr_t address;
    size_t size;
    hooks::Allocator allocator;
};
```
I'm wondering why error is called (unknown location)?

**Comment by godlygeek** [comment]
You already asked this in https://github.com/bloomberg/memray/discussions/629 and I answered it there. You're trying to use a library built for Python 3.9 with a 3.11 interpreter. That will not work. You need to use Python 3.11 to install the package in the `build-parent` container.

> I'm wondering why error is called (unknown location)?

It's telling you that it went looking for a module named `memray._memray` and couldn't find it - there was no file on disk matching one of the possible names for a `memray._memray` module that can be loaded by a 3.11 interpreter.

**Reply by godlygeek** [reply]
Ah, wait, nevermind, I misread - if you did switch to using 3.9 inside the distroless container since #629, then it's not just a version mismatch on the artifact that you're copying in. What do you see if you:
```sh
ls -l /venv/lib/python3.9/site-packages/memray
```
?

**Reply by Partho** [reply]
The distroless doesn't have bash but it has sh, hence I ran `echo *` at the location you mentioned 
(`/venv/lib/python3.9/site-packages/memray`) and I see this output:
```text
__init__.py 
__init__.pyi 
__main__.py 
__pycache__ 
_destination.py 
_errors.py 
_inject.abi3.so 
_ipython 
_memray 
_memray.cpython-39-x86_64-linux-gnu.so 
_memray.pyi 
_memray.pyx 
_memray_test_utils.pyx 
_metadata.py 
_stats.py 
_stats.pyi 
_test.py 
_test_utils.cpython-39-x86_64-linux-gnu.so 
_test_utils.pyi 
_version.py 
commands 
py.typed 
reporters
```

**Reply by godlygeek** [reply]
Hm, at a glance that looks reasonable. What do you see if you run these commands?
```sh
/venv/bin/python --version
/venv/bin/python -m site
/venv/bin/python -c "import importlib.machinery; print(importlib.machinery.all_suffixes())"
```

**Reply by Partho** [reply]
I got the following output:
```shell
# /venv/bin/python --version
Python 3.9.2

# /venv/bin/python -m site
sys.path = [
    '/',
    '/usr/lib/python39.zip',
    '/usr/lib/python3.9',
    '/usr/lib/python3.9/lib-dynload',
    '/venv/lib/python3.9/site-packages',
]
USER_BASE: '/root/.local' (doesn't exist)
USER_SITE: '/root/.local/lib/python3.9/site-packages' (doesn't exist)
ENABLE_USER_SITE: False

# /venv/bin/python -c "import importlib.machinery; print(importlib.machinery.all_suffixes())"
['.py', '.pyc', '.cpython-39-aarch64-linux-gnu.so', '.abi3.so', '.so']
```

**Reply by godlygeek** [reply, marked as answer]
I see: your `_memray.*.so` is called `_memray.cpython-39-x86_64-linux-gnu.so`, but your Python is looking for `_memray.cpython-39-aarch64-linux-gnu.so`. It appears that you copied x86-64 executable code from your base image into an aarch64 distroless image. That's why it's not working: an aarch64 (ARM) Python can't import modules built for x86_64 (Intel).
